### PR TITLE
Add slither for l1 contracts

### PR DESCRIPTION
**Pull request by dnkolegov**
# What ❔

This PR enables the slither static analyzer for L1 contracts and also resolves the issues it has found.
It is configured to catch high- and medium-severity potential issues.
It doesn't check `Verifier.sol` contract.

The PR also documents how to deal with findings.

## Why ❔
To find "low-hanging fruits" bugs and known vulnerabilities.

## Checklist

- [x] PR title corresponds to the body of PR (we generate changelog entries from PRs).
- [x] Tests for the changes have been added / updated.
- [x] Documentation comments have been added / updated.

### CODE REVIEW

**Comment by dnkolegov** [comment]
@StanislavBreadless as we discussed with @vladbochok could you please review the following findings

```
MailboxFacet._deriveL2GasPrice(uint256,uint256) (contracts/state-transition/chain-deps/facets/Mailbox.sol#163-182) performs a multiplication on the result of a division:
	- l1GasPriceConverted = (_l1GasPrice * s.baseTokenGasPriceMultiplierNominator) / s.baseTokenGasPriceMultiplierDenominator (contracts/state-transition/chain-deps/facets/Mailbox.sol#[16](https://github.com/matter-labs/era-contracts/actions/runs/8425289708/job/23071021468?pr=280#step:9:17)6-167)
	- batchOverheadBaseToken = uint256(feeParams.batchOverheadL1Gas) * l1GasPriceConverted (contracts/state-transition/chain-deps/facets/Mailbox.sol#173)
MailboxFacet._deriveL2GasPrice(uint256,uint256) (contracts/state-transition/chain-deps/facets/Mailbox.sol#163-182) performs a multiplication on the result of a division:
	- l1GasPriceConverted = (_l1GasPrice * s.baseTokenGasPriceMultiplierNominator) / s.baseTokenGasPriceMultiplierDenominator (contracts/state-transition/chain-deps/facets/Mailbox.sol#166-167)
	- pubdataPriceBaseToken = L1_GAS_PER_PUBDATA_BYTE * l1GasPriceConverted (contracts/state-transition/chain-deps/facets/Mailbox.sol#[17](https://github.com/matter-labs/era-contracts/actions/runs/8425289708/job/23071021468?pr=280#step:9:18)0)
Reference: https://github.com/crytic/slither/wiki/Detector-Documentation#divide-before-multiply
```
and let us know how we should proceed. Should it be fixed or ignored?  If fixed, then could you add a fix into this PR?
Thank you.

File: l1-contracts/foundry.toml [hunk 1]
```diff
@@ -1,6 +1,6 @@
 [profile.default]
 src = 'contracts'
-out = 'artifacts-forge'
+out = "out"
```
Comment (StanislavBreadless) [inline l1-contracts/foundry.toml:3]: why do we need this change?

**Comment by StanislavBreadless** [comment]
@dnkolegov we can ignore it. If we used what slither proposed, the calculations would indeed be slightly more precise, but IMHO the way it is now is a bit easier to read & reuse (i.e. whenever we need l1 gas price we can use `l1GasPriceConverted`)

File: .github/workflows/slither.yaml [hunk 2]
```diff
@@ -0,0 +1,43 @@
...
+      - name: Install Slither
+        run: |
+          pip install slither-analyzer
+
+      - name: Use Foundry
+        uses: foundry-rs/foundry-toolchain@v1
+
+      - name: Remove non-compiled files
+        run: rm -rf ./l1-contracts/contracts/state-transition/utils/ |
+          rm -rf ./l1-contracts/contracts/state-transition/Verifier.sol |
```
[the remark this answers is not in the document]
Comment (dnkolegov) [reply .github/workflows/slither.yaml:39]: Slither can't parse these contracts. I created a [bug](https://github.com/crytic/slither/issues/2369).

File: l1-contracts/foundry.toml (hunk 1 above)
Comment (dnkolegov) [reply l1-contracts/foundry.toml]: Unfortunately, slither does not work with `out = 'artifacts-forge'` even if to specify it explicitly via `--foundry-out-directory ./l1-contracts/artifacts-forge`

**Review by StanislavBreadless — APPROVED** [review]

**Pull request merged**